### PR TITLE
Add support for Cilium network policies

### PR DESCRIPTION
**Pull request by juliogreff**
### What does this PR do?

Initially introduced in the helm chart:
https://github.com/DataDog/helm-charts/pull/99/ -- this PR adds the same
policies and also includes a couple of fixes introduced by
https://github.com/DataDog/helm-charts/pull/245 and DataDog/helm-charts#387.

### Describe your test plan

Enable the cilium network policies for the agent, check that they are created correctly, and that agent, cluster agent, and cluster checks runners still work as expected, look out for networking errors in particular.

Follow these docs to install Cilium on a Kind cluster: https://docs.cilium.io/en/v1.9/gettingstarted/kind/

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -191,6 +191,8 @@ spec:
 | agent.log.podLogsPath | Allows log collection from pod log path. Defaults to `/var/log/pods`. |
 | agent.log.tempStoragePath | This path (always mounted from the host) is used by Datadog Agent to store information about processed log files. If the Datadog Agent is restarted, it starts tailing the log files immediately. Default to `/var/lib/datadog-agent/logs` |
 | agent.networkPolicy.create | If true, create a NetworkPolicy for the current agent. |
+| agent.networkPolicy.dnsSelectorEndpoints |  |
+| agent.networkPolicy.flavor | Flavor of the network policy to use. Can be `kubernetes` or `cilium`. |
```
Comment (kayayarai) [inline docs/configuration.md:195]:
```suggestion
| agent.networkPolicy.flavor | Which network policy to use. Can be `kubernetes` or `cilium`. |
```

File: docs/configuration.md [hunk 2]
```diff
@@ -191,6 +191,8 @@ spec:
 | agent.log.podLogsPath | Allows log collection from pod log path. Defaults to `/var/log/pods`. |
 | agent.log.tempStoragePath | This path (always mounted from the host) is used by Datadog Agent to store information about processed log files. If the Datadog Agent is restarted, it starts tailing the log files immediately. Default to `/var/lib/datadog-agent/logs` |
 | agent.networkPolicy.create | If true, create a NetworkPolicy for the current agent. |
+| agent.networkPolicy.dnsSelectorEndpoints |  |
```
Comment (kayayarai) [inline docs/configuration.md:194]: Needs a description.

File: docs/configuration.md [hunk 3]
```diff
@@ -305,6 +307,8 @@ spec:
 | clusterAgent.keepAnnotations | KeepAnnotations allows the specification of annotations not managed by the Operator that will be kept on ClusterAgent Deployment. All annotations containing 'datadoghq.com' are always included. This field uses glob syntax. |
 | clusterAgent.keepLabels | KeepLabels allows the specification of labels not managed by the Operator that will be kept on ClusterAgent Deployment. All labels containing 'datadoghq.com' are always included. This field uses glob syntax. |
 | clusterAgent.networkPolicy.create | If true, create a NetworkPolicy for the current agent. |
+| clusterAgent.networkPolicy.dnsSelectorEndpoints |  |
+| clusterAgent.networkPolicy.flavor | Flavor of the network policy to use. Can be `kubernetes` or `cilium`. |
```
Comment (kayayarai) [inline docs/configuration.md:311]: Same comments as above

File: docs/configuration.md [hunk 4]
```diff
@@ -365,6 +369,8 @@ spec:
 | clusterChecksRunner.image.pullSecrets | It is possible to specify docker registry credentials. See https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod |
 | clusterChecksRunner.image.tag | Define the image version to use: To be used if the Name field does not correspond to a full image string. |
 | clusterChecksRunner.networkPolicy.create | If true, create a NetworkPolicy for the current agent. |
+| clusterChecksRunner.networkPolicy.dnsSelectorEndpoints |  |
+| clusterChecksRunner.networkPolicy.flavor | Flavor of the network policy to use. Can be `kubernetes` or `cilium`. |
```
Comment (kayayarai) [inline docs/configuration.md:373]: same comments as above.

**Review by kayayarai — CHANGES_REQUESTED** [review]

File: controllers/datadogagent/agent.go [hunk 5]
```diff
@@ -394,6 +407,198 @@ func buildAgentNetworkPolicy(dda *datadoghqv1alpha1.DatadogAgent, name string) *
...
+								{
+									Port:     "51678",
+									Protocol: cilium.ProtocolTCP,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
```
Comment (L3n41c) [inline controllers/datadogagent/agent.go:474]: We might need to retrofit here DataDog/helm-charts#391 and to add
```yaml
ToEntities: []cilium.Entity{
	cilium.EntityHost,
},
```
to fix DataDog/helm-charts#387 also with the operator.

**Review by kayayarai — APPROVED** [review]

**Review by clamoriniere — COMMENTED** [review]
few comments
btw, I like the `builder` pattern 👍

File: pkg/cilium/types.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
+package cilium
```
[the remark this answers is not in the document]
Comment (juliogreff) [reply pkg/cilium/types.go]: Updated 👍

**Review by clamoriniere — APPROVED** [review]
Nice work 👍 

Do you have any test tips to enable cilium in a k8s cluster (kind, minikube? ) that can be added in the PR description?

**Pull request merged**